src/ripple/net/impl/HTTPStream.cpp

```diff
@@ -0,0 +1,191 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/net/HTTPStream.h>
+
+namespace ripple {
+
+SSLStream::SSLStream(
+    Config const& config,
+    boost::asio::io_service::strand& strand,
+    beast::Journal j)
+    : ssl_ctx_(config, j, boost::asio::ssl::context::tlsv12_client)
+    , strand_(strand)
+{
+}
+
+boost::asio::ip::tcp::socket&
+SSLStream::getStream()
+{
+    assert(stream_);
+    return stream_->next_layer();
+}
+
+bool
+SSLStream::connect(
+    std::string& errorOut,
+    std::string const host,
+    std::string const port,
+    boost::asio::yield_context& yield)
```

Comment:
We've lost `ec` error message loging. Consider passing in an `ec` ref as done in the other functions.

Reply:
Passing the error code to the write and read functions is easy because in each case, there's only one function that could be the cause of an error, whereas in the connect functions, there are many functions that could cause an error. So to provide info to the caller about which function errored, we'd have to provide additional context with the `ec` parameter. Instead of that I'd rather just provide the context (`errorOut`) without the additional `ec` parameter.

Reply:
While reporting "which function errored" may be helpful, we are throwing away the cause. Typically if the `ec` is set, further progress on the process is aborted, the call stack is unwind and the `ec` is floated up. If we choose to keep things as they, it may be helpful to append its message into `errorOut`. 

eg.
```return fail("async_resolve. Error: " + ec.message());```

As is, if any of these functions fail, we don't know why.

Reply:
For example, if the call to `resolver.async_resolve` fails, `errorOut` will contain "async_resolve" but the reason it failed will be discarded and we will have no idea why it failed.

Reply:
Done